src/palindrom.js

```diff
@@ -4,1052 +4,391 @@
+                this.onStateReset(this.obj);
+            }
+            this.onRemoteChange({ patches: sequence, results })
+        } catch (error) {
             if (this.debug) {
-                this.validateSequence(this.remoteObj, patches);
+                this.onIncomingPatchValidationError(error);
+                return;
+            } else {
+                throw error;
             }
-            sendPatches(this, this.queue.send(patches));
-            this.onLocalChange(patches);
         }
+        return this.obj;
+    }
 
-        validateAndApplySequence(tree, sequence) {
-            // we don't want this changes to generate patches since they originate from server, not client
-            try {
-                this.unobserve();
-                const results = applyPatch(tree, sequence, this.debug);
-                // notifications have to happen only where observe has been re-enabled
-                // otherwise some listener might produce changes that would go unnoticed
-                this.observe();
-                // the state was fully replaced
-                if (results.newDocument !== tree) {
-                    // object was reset, proxify it again
-                    this.prepareProxifiedObject(results.newDocument);
+    validateSequence(tree, sequence) {
+        const error = validate(sequence, tree);
+        if (error) {
+            this.onOutgoingPatchValidationError(error);
+        }
+    }
 
-                    this.queue.obj = this.obj;
+    /**
+     * Handle an error which is probably caused by random disconnection
+     */
+    handleConnectionError() {
+        this.heartbeat.stop();
+        this.reconnector.triggerReconnection();
+        this.onConnectionError();
+    }
 
-                    // validate json response
-                    findRangeErrors(
-                        this.obj,
-                        this.onIncomingPatchValidationError
-                    );
+    /**
+     * Handle an error which probably won't go away on itself (basically forward upstream)
+     * @param {PalindromConnectionError} palindromError
+     */
+    handleFatalError(palindromError) {
+        this.heartbeat.stop();
+        this.reconnector.stopReconnecting();
+        this.onConnectionError(palindromError);
+    }
 
-                    //notify people about it
-                    try {
-                        this.onStateReset(this.obj);
-                    } catch (error) {
-                        // to prevent the promise's catch from swallowing errors inside onStateReset
-                        this.onError(
-                            new PalindromError(
-                                `Error inside onStateReset callback: ${
-                                    error.message
-                                }`
-                            )
-                        );
-                        console.error(error);
```

Comment:
This seems to be not addressed. @tomalec what do we do about it?

Reply:
I guess explanation why it was there, is:
```
 // to prevent the promise's catch from swallowing errors inside onStateReset
```
Is it needed any longer. Well from green tests, I'd say "no" :trollface:, then once it will hit us back, we would have to write test for it.

Reply:
Test didn't catch it, because the one testing it was removed https://github.com/Palindrom/Palindrom/pull/227/files/87602127cfaf7d5b9c19b2b5f51b996d8201b033#r287282209

Reply:
Restored the functionality and the test.